File: python/cldpm/commands/clone.py
"""Implementation of cldpm clone command."""

import json
import shutil
from pathlib import Path

import click

from ..core.config import load_cldpm_config, load_project_config
from ..core.resolver import resolve_project
from ..utils.fs import ensure_dir, find_repo_root
from ..utils.output import console, print_error, print_success, print_dir_tree


@click.command()
@click.argument("project_name")
@click.argument("directory")
@click.option(
    "--include-shared",
    is_flag=True,
    help="Also copy the full shared/ directory structure",
)
@click.option(
    "--preserve-links",
    is_flag=True,
    help="Keep symlinks instead of copying (requires shared/ to exist)",
)
def clone(
    project_name: str,
    directory: str,
    include_shared: bool,
    preserve_links: bool,
) -> None:
    """Clone a project to a standalone directory with all dependencies.

    Creates a complete, standalone copy of the project:
      - Shared components: Resolved from symlinks, copied as actual files
      - Local components: Copied directly

    \b
    Use cases:
      - Export project to work outside the mono repo
      - Share project with someone without the mono repo
      - Create a snapshot with all dependencies

    \b
    Examples:
      cldpm clone my-project ./standalone
      cldpm clone my-project /path/to/output
      cldpm clone my-project ./export --include-shared
      cldpm clone my-project ./linked --preserve-links
    """
    # Find repo root
    repo_root = find_repo_root()
    if repo_root is None:
        print_error("Not in a CLDPM mono repo. Run 'cldpm init' first.")
        raise SystemExit(1)

    # Resolve project
    try:
        resolved = resolve_project(project_name, repo_root)
    except FileNotFoundError as e:
        print_error(str(e))
        raise SystemExit(1)

    # Get source project path
    source_path = Path(resolved["path"])
    target_path = Path(directory).resolve()

    # Check if target exists
    if target_path.exists():
        print_error(f"Target directory already exists: {target_path}")
        raise SystemExit(1)

    # Create target directory
    ensure_dir(target_path)

    # Copy project files (excluding .claude subdirectories that are symlinks)
    for item in source_path.iterdir():
        dest = target_path / item.name

        if item.name == ".claude":
            # Handle .claude directory specially
            ensure_dir(dest)
            for claude_item in item.iterdir():
                claude_dest = dest / claude_item.name

                if claude_item.name in ["skills", "agents", "hooks", "rules"]:
                    # Create directory, will be populated below
                    ensure_dir(claude_dest)

                    # Copy local (non-symlink) components directly
                    for comp_item in claude_item.iterdir():
                        if comp_item.name == ".gitignore":
                            continue  # Skip .gitignore, will regenerate if needed
                        if not comp_item.is_symlink():
                            comp_dest = claude_dest / comp_item.name
                            if comp_item.is_dir():
                                shutil.copytree(comp_item, comp_dest)
                            else:
                                shutil.copy2(comp_item, comp_dest)
                elif claude_item.is_file():
                    shutil.copy2(claude_item, claude_dest)
                elif claude_item.is_dir():
                    shutil.copytree(claude_item, claude_dest)
        elif item.is_dir():
            shutil.copytree(item, dest, symlinks=preserve_links)
        else:
            shutil.copy2(item, dest)

    # Copy shared dependencies
    cldpm_config = load_cldpm_config(repo_root)
    shared_dir = repo_root / cldpm_config.shared_dir

    shared_counts = {}
    for dep_type in ["skills", "agents", "hooks", "rules"]:
        count = 0
        for component in resolved["shared"].get(dep_type, []):
            comp_name = component["name"]
            source_comp = shared_dir / dep_type / comp_name
            target_comp = target_path / ".claude" / dep_type / comp_name

            if source_comp.exists() and not target_comp.exists():
                if preserve_links:
                    # Create symlink (will only work if shared/ is accessible)
                    target_comp.symlink_to(source_comp)
                else:
                    # Copy actual files
                    if source_comp.is_dir():
                        shutil.copytree(source_comp, target_comp)
                    else:
                        shutil.copy2(source_comp, target_comp)
                count += 1
        shared_counts[dep_type] = count

    # Count local components
    local_counts = {
        dep_type: len(resolved["local"].get(dep_type, []))
        for dep_type in ["skills", "agents", "hooks", "rules"]
    }

    # Optionally copy full shared directory
    if include_shared:
        target_shared = target_path / "shared"
        shutil.copytree(shared_dir, target_shared)

        # Also copy cldpm.json for reference
        shutil.copy2(repo_root / "cldpm.json", target_path / "cldpm.json")

    # Update settings.json if it references hooks
    settings_path = target_path / ".claude" / "settings.json"
    if settings_path.exists():
        try:
            with open(settings_path) as f:
                settings = json.load(f)

            # Update any hook paths to be relative to the cloned project
            # (This is a placeholder - actual hook path updates would depend on the hook format)

            with open(settings_path, "w") as f:
                json.dump(settings, f, indent=2)
                f.write("\n")
        except (json.JSONDecodeError, IOError):
            pass  # Ignore settings.json issues

    print_success(f"Cloned {project_name} to {target_path}")

    # Show what was copied
    non_zero_shared = {k: v for k, v in shared_counts.items() if v > 0}
    non_zero_local = {k: v for k, v in local_counts.items() if v > 0}

    if non_zero_shared:
        deps_str = ", ".join(f"{v} {k}" for k, v in non_zero_shared.items())
        console.print(f"  Shared: {deps_str}")

    if non_zero_local:
        deps_str = ", ".join(f"{v} {k}" for k, v in non_zero_local.items())
        console.print(f"  Local: {deps_str}")

    if include_shared:
        console.print("  Included: full shared/ directory")

    console.print()
    print_dir_tree(target_path, max_depth=3)

File: typescript/src/commands/clone.ts
/**
 * cldpm clone command
 */

import { Command } from "commander";
import { cwd } from "node:process";
import * as path from "node:path";
import * as fs from "node:fs/promises";
import { resolveProject, loadCldpmConfig } from "../core/index.js";
import { error, success } from "../utils/index.js";

/**
 * Copy a directory recursively
 */
async function copyDir(
  src: string,
  dest: string,
  options: { preserveSymlinks?: boolean } = {}
): Promise<void> {
  await fs.mkdir(dest, { recursive: true });
  const entries = await fs.readdir(src, { withFileTypes: true });

  for (const entry of entries) {
    const srcPath = path.join(src, entry.name);
    const destPath = path.join(dest, entry.name);

    // Check if it's a symlink using lstat
    const lstat = await fs.lstat(srcPath);

    if (lstat.isSymbolicLink()) {
      if (options.preserveSymlinks) {
        // Preserve symlink
        const linkTarget = await fs.readlink(srcPath);
        await fs.symlink(linkTarget, destPath);
      } else {
        // Resolve symlink and copy the actual content
        const realPath = await fs.realpath(srcPath);
        const stat = await fs.stat(realPath);
        if (stat.isDirectory()) {
          await copyDir(realPath, destPath, options);
        } else {
          await fs.copyFile(realPath, destPath);
        }
      }
    } else if (entry.isDirectory()) {
      await copyDir(srcPath, destPath, options);
    } else {
      await fs.copyFile(srcPath, destPath);
    }
  }
}

/**
 * Print a simple directory tree
 */
async function printDirTree(
  dirPath: string,
  prefix = "",
  maxDepth = 3,
  currentDepth = 0
): Promise<void> {
  if (currentDepth >= maxDepth) return;

  const entries = await fs.readdir(dirPath, { withFileTypes: true });
  const sortedEntries = entries.sort((a, b) => {
    // Directories first, then files
    if (a.isDirectory() && !b.isDirectory()) return -1;
    if (!a.isDirectory() && b.isDirectory()) return 1;
    return a.name.localeCompare(b.name);
  });

  for (let i = 0; i < sortedEntries.length; i++) {
    const entry = sortedEntries[i]!;
    const isLast = i === sortedEntries.length - 1;
    const connector = isLast ? "└── " : "├── ";
    const childPrefix = isLast ? "    " : "│   ";

    console.log(`${prefix}${connector}${entry.name}`);

    if (entry.isDirectory()) {
      await printDirTree(
        path.join(dirPath, entry.name),
        prefix + childPrefix,
        maxDepth,
        currentDepth + 1
      );
    }
  }
}

export const cloneCommand = new Command("clone")
  .description("Clone a project with all dependencies")
  .argument("<project>", "Project name")
  .argument("<directory>", "Target directory")
  .option("--include-shared", "Also copy the full shared/ directory structure")
  .option("--preserve-links", "Keep symlinks instead of copying")
  .action(
    async (
      projectName: string,
      directory: string,
      options: {
        includeShared?: boolean;
        preserveLinks?: boolean;
      }
    ) => {
      try {
        const repoRoot = cwd();

        // Resolve project
        const resolved = await resolveProject(projectName, repoRoot);

        // Get paths
        const sourcePath = resolved.path;
        const targetPath = path.resolve(directory);

        // Check if target exists
        try {
          await fs.access(targetPath);
          error(`Target directory already exists: ${targetPath}`);
          process.exit(1);
        } catch {
          // Directory doesn't exist, which is what we want
        }

        // Create target directory
        await fs.mkdir(targetPath, { recursive: true });

        // Load CLDPM config for shared directory path
        const cldpmConfig = await loadCldpmConfig(repoRoot);
        const sharedDir = path.join(repoRoot, cldpmConfig.sharedDir);

        // Copy project files
        const entries = await fs.readdir(sourcePath, { withFileTypes: true });
        for (const entry of entries) {
          const srcPath = path.join(sourcePath, entry.name);
          const destPath = path.join(targetPath, entry.name);

          if (entry.name === ".claude") {
            // Handle .claude directory specially
            await fs.mkdir(destPath, { recursive: true });
            const claudeEntries = await fs.readdir(srcPath, { withFileTypes: true });

            for (const claudeEntry of claudeEntries) {
              const claudeSrcPath = path.join(srcPath, claudeEntry.name);
              const claudeDestPath = path.join(destPath, claudeEntry.name);

              if (["skills", "agents", "hooks", "rules"].includes(claudeEntry.name)) {
                // Create directory
                await fs.mkdir(claudeDestPath, { recursive: true });

                // Copy local (non-symlink) components directly
                const compEntries = await fs.readdir(claudeSrcPath, { withFileTypes: true });
                for (const compEntry of compEntries) {
                  if (compEntry.name === ".gitignore") continue;

                  const compSrcPath = path.join(claudeSrcPath, compEntry.name);
                  const compDestPath = path.join(claudeDestPath, compEntry.name);

                  // Check if it's a symlink
                  const lstat = await fs.lstat(compSrcPath);
                  if (!lstat.isSymbolicLink()) {
                    if (compEntry.isDirectory()) {
                      await copyDir(compSrcPath, compDestPath, {
                        preserveSymlinks: options.preserveLinks,
                      });
                    } else {
                      await fs.copyFile(compSrcPath, compDestPath);
                    }
                  }
                }
              } else if (claudeEntry.isFile()) {
                await fs.copyFile(claudeSrcPath, claudeDestPath);
              } else if (claudeEntry.isDirectory()) {
                await copyDir(claudeSrcPath, claudeDestPath, {
                  preserveSymlinks: options.preserveLinks,
                });
              }
            }
          } else if (entry.isDirectory()) {
            await copyDir(srcPath, destPath, {
              preserveSymlinks: options.preserveLinks,
            });
          } else {
            await fs.copyFile(srcPath, destPath);
          }
        }

        // Copy shared dependencies
        const depTypes = ["skills", "agents", "hooks", "rules"] as const;
        const sharedCounts: Record<string, number> = {};

        for (const depType of depTypes) {
          let count = 0;
          const components = resolved.shared[depType] || [];

          for (const component of components) {
            const compName = component.name;
            const sourceComp = path.join(sharedDir, depType, compName);
            const targetComp = path.join(targetPath, ".claude", depType, compName);

            try {
              await fs.access(sourceComp);
              try {
                await fs.access(targetComp);
                // Already exists, skip
              } catch {
                // Doesn't exist, copy or link
                if (options.preserveLinks) {
                  // Create symlink
                  await fs.symlink(sourceComp, targetComp);
                } else {
                  // Copy actual files
                  const stat = await fs.stat(sourceComp);
                  if (stat.isDirectory()) {
                    await copyDir(sourceComp, targetComp);
                  } else {
                    await fs.copyFile(sourceComp, targetComp);
                  }
                }
                count++;
              }
            } catch {
              // Source doesn't exist, skip
            }
          }
          sharedCounts[depType] = count;
        }

        // Count local components
        const localCounts: Record<string, number> = {};
        for (const depType of depTypes) {
          localCounts[depType] = (resolved.local[depType] || []).length;
        }

        // Optionally copy full shared directory
        if (options.includeShared) {
          const targetShared = path.join(targetPath, "shared");
          await copyDir(sharedDir, targetShared);

          // Also copy cldpm.json
          await fs.copyFile(
            path.join(repoRoot, "cldpm.json"),
            path.join(targetPath, "cldpm.json")
          );
        }

        // Update settings.json if it exists (placeholder for hook path updates)
        const settingsPath = path.join(targetPath, ".claude", "settings.json");
        try {
          await fs.access(settingsPath);
          const content = await fs.readFile(settingsPath, "utf-8");
          const settings = JSON.parse(content);
          // Placeholder for any settings updates
          await fs.writeFile(settingsPath, JSON.stringify(settings, null, 2) + "\n");
        } catch {
          // No settings.json or invalid, skip
        }

        success(`Cloned ${projectName} to ${targetPath}`);

        // Show what was copied
        const nonZeroShared = Object.entries(sharedCounts).filter(([, v]) => v > 0);
        const nonZeroLocal = Object.entries(localCounts).filter(([, v]) => v > 0);

        if (nonZeroShared.length > 0) {
          const depsStr = nonZeroShared.map(([k, v]) => `${v} ${k}`).join(", ");
          console.log(`  Shared: ${depsStr}`);
        }

        if (nonZeroLocal.length > 0) {
          const depsStr = nonZeroLocal.map(([k, v]) => `${v} ${k}`).join(", ");
          console.log(`  Local: ${depsStr}`);
        }

        if (options.includeShared) {
          console.log("  Included: full shared/ directory");
        }

        console.log();
        console.log(path.basename(targetPath));
        await printDirTree(targetPath);
      } catch (err) {
        error(`Failed to clone project: ${(err as Error).message}`);
        process.exit(1);
      }
    }
  );
